xbmc/settings/SettingsValueFlatJsonSerializer.cpp

```diff
@@ -0,0 +1,130 @@
+  {
+    CLog::Log(LOGWARNING, "CSettingsValueFlatJsonSerializer: failed to serialize settings into JSON");
+    return "";
+  }
+
+  return result;
+}
+
+void CSettingsValueFlatJsonSerializer::SerializeSection(CVariant& parent, std::shared_ptr<CSettingSection> section) const
+{
+  if (section == nullptr)
+    return;
+
+  const auto categories = section->GetCategories();
+  for (const auto& category : categories)
+    SerializeCategory(parent, category);
+}
+
+void CSettingsValueFlatJsonSerializer::SerializeCategory(CVariant& parent, std::shared_ptr<CSettingCategory> category) const
+{
+  if (category == nullptr)
+    return;
+
+  const auto groups = category->GetGroups();
+  for (const auto& group : groups)
+    SerializeGroup(parent, group);
+}
+
+void CSettingsValueFlatJsonSerializer::SerializeGroup(CVariant& parent, std::shared_ptr<CSettingGroup> group) const
+{
+  if (group == nullptr)
+    return;
+
+  const auto settings = group->GetSettings();
+  for (const auto& setting : settings)
+    SerializeSetting(parent, setting);
+}
+
+void CSettingsValueFlatJsonSerializer::SerializeSetting(CVariant& parent, std::shared_ptr<CSetting> setting) const
+{
+  if (setting == nullptr)
+    return;
+
+  // ignore references and action settings (which don't have a value)
+  if (setting->IsReference() || setting->GetType() == SettingType::Action)
+    return;
+
+  const auto valueObj = SerializeSettingValue(setting);
+  if (valueObj.isNull())
+    return;
+
+  parent[setting->GetId()] = valueObj;
+}
+
+CVariant CSettingsValueFlatJsonSerializer::SerializeSettingValue(std::shared_ptr<CSetting> setting) const
+{
+  switch (setting->GetType())
+  {
+  case SettingType::Action:
+    return CVariant::ConstNullVariant;
+
+  case SettingType::Boolean:
+    return CVariant(std::static_pointer_cast<CSettingBool>(setting)->GetValue());
+
+  case SettingType::Integer:
+    return CVariant(std::static_pointer_cast<CSettingInt>(setting)->GetValue());
+
+  case SettingType::Number:
+    return CVariant(std::static_pointer_cast<CSettingNumber>(setting)->GetValue());
+
+  case SettingType::String:
+    return CVariant(std::static_pointer_cast<CSettingString>(setting)->GetValue());
+
+  case SettingType::List:
+  {
+    const auto settingList = std::static_pointer_cast<CSettingList>(setting);
+
+    CVariant settingListValuesObj(CVariant::VariantTypeArray);
+    const auto settingListValues = settingList->GetValue();
+    for (const auto& settingListValue : settingListValues)
+    {
+      const auto valueObj = SerializeSettingValue(settingListValue);
+      if (!valueObj.isNull())
+        settingListValuesObj.push_back(valueObj);
+    }
+
+    return settingListValuesObj;
+  }
+
+  case SettingType::Unknown:
+  default:
+    CLog::Log(LOGWARNING, "CSettingsValueFlatJsonSerializer: failed to serialize setting \"{}\" with value \"{}\" of unknown type", setting->GetId(), setting->ToString());
+    return CVariant::ConstNullVariant;
```

Comment:
> Use spaces as tab policy with an indentation size of 2.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

The no indentation exception for switches was dropped recently (#16701).

Reply:
Fixed.